pkg/server/commands/batch_check_command.go

```diff
@@ -0,0 +1,162 @@
+package commands
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/openfga/openfga/internal/concurrency"
+	"github.com/openfga/openfga/internal/server/config"
+
+	"github.com/openfga/openfga/pkg/server/errors"
+
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+
+	"github.com/openfga/openfga/internal/cachecontroller"
+	"github.com/openfga/openfga/internal/graph"
+	"github.com/openfga/openfga/pkg/logger"
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/openfga/openfga/pkg/typesystem"
+)
+
+type BatchCheckQuery struct {
+	cacheController     cachecontroller.CacheController
+	checkResolver       graph.CheckResolver
+	datastore           storage.RelationshipTupleReader
+	logger              logger.Logger
+	maxChecksAllowed    uint32
+	maxConcurrentChecks uint32
+	typesys             *typesystem.TypeSystem
+}
+
+type BatchCheckCommandParams struct {
+	AuthorizationModelID string
+	Checks               []*openfgav1.BatchCheckItem
+	Consistency          openfgav1.ConsistencyPreference
+	StoreID              string
+}
+
+type BatchCheckOutcome struct {
+	CheckResponse *graph.ResolveCheckResponse
+	Duration      time.Duration
+	Err           error
+}
+
+type BatchCheckQueryOption func(*BatchCheckQuery)
+
+func WithBatchCheckCommandCacheController(cc cachecontroller.CacheController) BatchCheckQueryOption {
+	return func(bq *BatchCheckQuery) {
+		bq.cacheController = cc
+	}
+}
+
+func WithBatchCheckCommandLogger(l logger.Logger) BatchCheckQueryOption {
+	return func(bq *BatchCheckQuery) {
+		bq.logger = l
+	}
+}
+
+func WithBatchCheckMaxConcurrentChecks(maxConcurrentChecks uint32) BatchCheckQueryOption {
+	return func(bq *BatchCheckQuery) {
+		bq.maxConcurrentChecks = maxConcurrentChecks
+	}
+}
+
+func WithBatchCheckMaxChecksPerBatch(maxChecks uint32) BatchCheckQueryOption {
+	return func(bq *BatchCheckQuery) {
+		bq.maxChecksAllowed = maxChecks
+	}
+}
+
+func NewBatchCheckCommand(datastore storage.RelationshipTupleReader, checkResolver graph.CheckResolver, typesys *typesystem.TypeSystem, opts ...BatchCheckQueryOption) *BatchCheckQuery {
+	cmd := &BatchCheckQuery{
+		logger:              logger.NewNoopLogger(),
+		datastore:           datastore,
+		cacheController:     cachecontroller.NewNoopCacheController(),
+		checkResolver:       checkResolver,
+		typesys:             typesys,
+		maxChecksAllowed:    config.DefaultMaxChecksPerBatchCheck,
+		maxConcurrentChecks: config.DefaultMaxConcurrentChecksPerBatchCheck,
+	}
+
+	for _, opt := range opts {
+		opt(cmd)
+	}
+	return cmd
+}
+
+func (bq *BatchCheckQuery) Execute(ctx context.Context, params *BatchCheckCommandParams) (map[string]*BatchCheckOutcome, error) {
+	if len(params.Checks) > int(bq.maxChecksAllowed) {
+		return nil, errors.ValidationError(
+			fmt.Errorf("batchCheck received %d checks, the maximum allowed is %d ", len(params.Checks), bq.maxChecksAllowed),
+		)
+	}
+
+	if err := validateNoDuplicateCorrelationIDs(params.Checks); err != nil {
+		return nil, err
```

Comment:
**Question:** Additional validations need to be made on the correlation IDs, right? For instance, making sure they're not empty strings?

Reply:
Correct, but those should be handled by [this rule in the API layer](https://github.com/openfga/api/pull/203/files#diff-d3a74ea9ecaac0a46611253dc0c4bc9c402e2ad9eb255cd41d07266453e258f0R1278)

Reply:
Keep in mind that OpenFGA can be interfaced with as a standalone package as opposed to via an API. In these cases, the API-layer validations won't be enforced. This may require some redundant checks and unit tests to ensure proper validation for each respective scenario.

Reply:
ah that is a great point, will add the needed validation